src/video/video.c

```diff
@@ -75,6 +71,116 @@ bool create_gl_context() {
+    // This must be loaded after we have the GL context up.
+    GLuint err = glewInit();
+    if(err != GLEW_OK) {
+        PERROR("Failed to load GLEW: %s", glewGetErrorString(err));
+        return false;
+    }
+    INFO("Loaded GLEW %s", glewGetString(GLEW_VERSION));
+    return true;
+}
+
+void print_log(const char *buffer, long len, const char *header) {
+    str log;
+    str sub;
+
+    // Make sure the string always ends in one \n
+    str_from_buf(&log, buffer, len);
+    str_rstrip(&log);
+    str_append_c(&log, "\n");
+
+    // Print line by line
+    size_t pos = 0, last = 0;
+    PERROR("--- %s ---", header);
+    while(str_find_next(&log, '\n', &pos)) {
+        str_from_slice(&sub, &log, last, pos - 1);
+        PERROR("%s", str_c(&sub));
+        str_free(&sub);
+        last = ++pos;
+    }
+    PERROR("--- end log ---", header);
+
+    str_free(&log);
+}
+
+void print_shader_log(const GLuint shader, const char *header) {
+    if(!glIsShader(shader)) {
+        PERROR("Attempted to print logs for shader %d: not a shader", shader);
+        return;
+    }
+
+    int buffer_size = 0, read_size = 0;
+    glGetShaderiv(shader, GL_INFO_LOG_LENGTH, &buffer_size);
+    char *buffer_data = omf_calloc(buffer_size, 1);
+    glGetShaderInfoLog(shader, buffer_size, &read_size, buffer_data);
+    print_log(buffer_data, read_size, header);
+    omf_free(buffer_data);
+}
+
+void print_program_log(const GLuint program) {
+    if(!glIsProgram(program)) {
+        PERROR("Attempted to print logs for program %d: not a program", program);
+        return;
+    }
+
+    int buffer_size = 0, read_size = 0;
+    glGetProgramiv(program, GL_INFO_LOG_LENGTH, &buffer_size);
+    char *buffer_data = omf_calloc(buffer_size, 1);
+    glGetProgramInfoLog(program, buffer_size, &read_size, buffer_data);
+    print_log(buffer_data, read_size, "program");
+    omf_free(buffer_data);
+}
+
+bool load_shader(GLuint program_id, GLenum shader_type, const char *shader_file) {
+    str shader_source;
+    str_from_file(&shader_source, shader_file);
+    const char *c_str = str_c(&shader_source);
+
+    GLuint shader = glCreateShader(shader_type);
+    glShaderSource(shader, 1, &c_str, NULL);
+    glCompileShader(shader);
+
+    str_free(&shader_source);
+
+    GLint status = GL_FALSE;
+    glGetShaderiv(shader, GL_COMPILE_STATUS, &status);
+    if(status != GL_TRUE) {
+        PERROR("Compilation error for shader %d (file=%s)", shader, shader_file);
+        print_shader_log(shader, shader_file);
+        return false;
+    }
+    DEBUG("Compilation succeeded for shader %d (file=%s)", shader, shader_file);
+    glAttachShader(program_id, shader);
+    return true;
+}
+
+bool create_program(GLuint *program_id, const char *vertex_shader, const char *fragment_shader) {
+    GLuint id = glCreateProgram();
+    if(!load_shader(id, GL_VERTEX_SHADER, vertex_shader))
+        goto error_0;
+    if(!load_shader(id, GL_FRAGMENT_SHADER, fragment_shader))
+        goto error_0;
```

Comment:
This leaks vertex shader if fragment shader fails.

Reply:
pfft, as if it would ever fail.


... right ?

Reply:
(fixed)